app/src/main/java/io/gnosis/safe/ui/transaction/TransactionsViewModel.kt

```diff
@@ -1,103 +1,122 @@
-                    ?: ViewAction.ShowEmptyState
-            )
-        }
-    }
-
-    private fun mapTransactions(
-        transactions: Page<Transaction>,
-        safe: Solidity.Address
-    ): List<TransactionView> {
-        return transactions.results.mapNotNull { transaction ->
-            when (transaction) {
-                is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
-                is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
-                else -> null
+        currentSafeAddress = safe
+
+        val safeTxItems: Flow<PagingData<TransactionView>> = transactionsPagingRepository.getTransactionsStream(safe)
+            .map { pagingData ->
+                pagingData.map { transaction ->
+                    when (transaction) {
+                        is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
+                        is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
+                        else -> TransactionView.Unknown
+                    }
+                }.filter { it !is TransactionView.Unknown }
             }
-        }
+            .map {
+                // insert headers
+                it.insertSeparators { before, after ->
+
+                    if (after == null) {
+                        // we're at the end of the list
+                        return@insertSeparators null  // no separator
+                    }
+
+                    if (before == null) {
+                        // we're at the beginning of the list
+                        return@insertSeparators when (after.transaction?.status) {
+                            TransactionStatus.Pending,
+                            TransactionStatus.AwaitingConfirmation,
+                            TransactionStatus.AwaitingExecution -> TransactionView.SectionHeader(title = R.string.tx_list_queue)
+
+                            TransactionStatus.Cancelled,
+                            TransactionStatus.Failed,
+                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)
+
+                            else -> null  // no separator
+                        }
+                    }
+
+                    when (before.transaction?.status) {
+                        TransactionStatus.Pending,
+                        TransactionStatus.AwaitingConfirmation,
+                        TransactionStatus.AwaitingExecution -> when (after.transaction?.status) {
+                            TransactionStatus.Cancelled,
+                            TransactionStatus.Failed,
+                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)
+                            else -> null  // no separator
+                        }
+                        else -> null  // no separator
```

Comment:
This looks a bit bulky, do you think we could split up this lambda into a couple of methods? To know exactly what's going on. 

All the nested `when` blocks make it a bit hard to read.

Reply:
here it is still clear what is going on. I think it is easier to read and understand now when everything is in one place. Should it become bulkier, more nested -> sure, I will dissect it into small pieces of easily digestible code

Reply:
I think that each `if` makes sense as a single method. Then you can describe with the method name what each `if` branch does.

Reply:
please provide code snippet how you see it

Reply:
So generally when you have 2 levels of when blocks I would suggest adding methods with names that explain your code somewhat. Like you see for the 2nd method, I don't really know how and why there is a need to add the `history` header again in certain cases. 

```kotlin
                it.insertSeparators { before, after ->

                    if (after == null) {
                        // we're at the end of the list
                        return@insertSeparators null  // no separator
                    }

// start of fun "setListHeader"
                    if (before == null) {
                        // we're at the beginning of the list
                        return@insertSeparators when (after.transaction?.status) {
                            TransactionStatus.Pending,
                            TransactionStatus.AwaitingConfirmation,
                            TransactionStatus.AwaitingExecution -> TransactionView.SectionHeader(title = R.string.tx_list_queue)

                            TransactionStatus.Cancelled,
                            TransactionStatus.Failed,
                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)

                            else -> null  // no separator
                        }
                    }
// end of fun "setListHeader"

// start of "iHaveNoIdeaWhatThisCodeDoes"
                    when (before.transaction?.status) {
                        TransactionStatus.Pending,
                        TransactionStatus.AwaitingConfirmation,
                        TransactionStatus.AwaitingExecution -> when (after.transaction?.status) {
                            TransactionStatus.Cancelled,
                            TransactionStatus.Failed,
                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)
                            else -> null  // no separator
                        }
                        else -> null  // no separator
               }
// end of "iHaveNoIdeaWhatThisCodeDoes"
        }
}
```